### PR TITLE
Support for J1939 requests on PGNs with and without payload

### PR DESCRIPTION
**Pull request by tjaehnel**

### CODE REVIEW

File: src/ecu_lua_script.cpp [hunk 1]
```diff
@@ -550,8 +555,33 @@ J1939PGNData EcuLuaScript::getJ1939PGNData(const string& pgn)
...
+        //offset for the first byte
+        int counter = 0;
+        while(val.exists() == false && pgnRequestPayloadWorking.length() < payloadTmp.length()){
+            //appends wildcard sign after the bytes that are tested
+            pgnRequestPayloadWorking = payloadTmp.substr(0,counter).append("*");
+            string pgnLookupWorking = pgn + " " + pgnRequestPayloadWorking;
+            cout << "Looking for: '" << pgnLookupWorking << "'" << endl;
+            val = lua_state_[ecu_ident_.c_str()][J1939_PGN_TABLE][pgnLookupWorking.c_str()];
+            //counter + blank + bytelength
+            counter = counter + 3;
```
Comment (marcr) [inline src/ecu_lua_script.cpp:570]: Why +3? Can you change that to a constant with a more meaningful name?

File: src/j1939_simulator.cpp [hunk 2]
```diff
@@ -213,47 +168,109 @@ int J1939Simulator::readData() noexcept
...
+    }
+    if(ackInfo.size() > 1) { // Group Function Value
+        rawMessage.push_back(ackInfo[1]);
+    } else {
+        rawMessage.push_back(0x00);
+    }
+    rawMessage.push_back(0xFF); // Reserved
+    rawMessage.push_back(0xFF); // Reserved
+    rawMessage.push_back(targetAddress); // Address Acknowledged
+    rawMessage.push_back((uint8_t)(pgn >> 0)); // PGN of requested information
```
Comment (marcr) [inline src/j1939_simulator.cpp:270]: I think you're reversing the order of the pgn bytes here, when pushing back and starting with the lowest byte, aren't you? Is that intentional?

File: src/j1939_simulator.cpp [hunk 3]
```diff
@@ -371,24 +388,33 @@ uint32_t J1939Simulator::parsePGN(string pgn) const noexcept
...
     addr.can_addr.j1939.pgn = J1939_NO_PGN; // Receive all PGNs
     addr.can_addr.j1939.name = J1939_NO_NAME; // Don't do name lookups
-    addr.can_addr.j1939.addr = source_address_; // source address
+    addr.can_addr.j1939.addr = node_address; // source address
 
     int skt = socket(PF_CAN, SOCK_DGRAM, CAN_J1939);
     if (skt < 0)
     {
-        cerr << __func__ << "() socket: " << strerror(errno) << endl;
+        cerr << __func__ << "() socket: " << strerror(errno) << '\n';
```
Comment (marcr) [inline src/j1939_simulator.cpp]: Out of curiosity, what was the issue with endl here?

File: src/ecu_lua_script.cpp [hunk 4]
```diff
@@ -620,3 +650,55 @@ void EcuLuaScript::registerIsoTpSender(IsoTpSender* pSender) noexcept
...
     pIsoTpSender_ = pSender;
 }
+
+string EcuLuaScript::intToHexString(const uint8_t* buffer, const size_t num_bytes)
+{
+    string a = "";
+
+    for (unsigned int i = 0; i < num_bytes; i++)
+    {
+        if ((buffer[i] / 16) > 9)
```
Comment (heinrichWeichert) [inline src/ecu_lua_script.cpp]: -> variable

File: src/ecu_lua_script.cpp [hunk 5]
```diff
@@ -620,3 +650,55 @@ void EcuLuaScript::registerIsoTpSender(IsoTpSender* pSender) noexcept
 {
     pIsoTpSender_ = pSender;
 }
+
+string EcuLuaScript::intToHexString(const uint8_t* buffer, const size_t num_bytes)
+{
+    string a = "";
+
+    for (unsigned int i = 0; i < num_bytes; i++)
```
Comment (heinrichWeichert) [inline src/ecu_lua_script.cpp]: https://stackoverflow.com/questions/26503606/better-way-to-convert-a-vector-of-uint8-to-an-ascii-hexadecimal-string

**Review by heinrichWeichert — CHANGES_REQUESTED** [review]

File: src/ecu_lua_script.cpp (hunk 5 above)
Comment (tobiasjaehnelavl) [reply src/ecu_lua_script.cpp]: Done

File: src/ecu_lua_script.cpp (hunk 4 above)
Comment (tobiasjaehnelavl) [reply src/ecu_lua_script.cpp]: I don't get it.
Is this still necessary?

**Review by heinrichWeichert — APPROVED** [review]

File: src/j1939_simulator.cpp (hunk 3 above)
Comment (heinrichWeichert) [reply src/j1939_simulator.cpp]: i guess endl is system-specific?

Comment (marcr) [reply src/j1939_simulator.cpp]: I first thought so, too, in regard to \r\n, but it's defined as \n only with additional flush at the end: https://cplusplus.com/reference/ostream/endl/
That's why I was wondering

File: src/ecu_lua_script.cpp (hunk 1 above)
Comment (tobiasjaehnelavl) [reply src/ecu_lua_script.cpp:570]: This code was already removed in https://github.com/AVL-DiTEST-DiagDev/car-simulator/pull/13/files#diff-8993c4e5b7566053240dc61431f9461a59e8ebdd8b7458d12baee6407290131cL558

File: src/j1939_simulator.cpp (hunk 2 above)
Comment (tobiasjaehnelavl) [reply src/j1939_simulator.cpp:270]: Yes, this is how J1939 works. ;-)

File: src/j1939_simulator.cpp (hunk 3 above)
Comment (tobiasjaehnelavl) [reply src/j1939_simulator.cpp]: I'm wondering how that changed.
I reverted that back to endl as this is the correct way to do it.

File: src/ecu_lua_script.cpp (hunk 1 above)
Comment (marcr) [reply src/ecu_lua_script.cpp:570]: Ok

File: src/j1939_simulator.cpp (hunk 2 above)
Comment (marcr) [reply src/j1939_simulator.cpp:270]: Ok :-)

**Review by marcr — APPROVED** [review]

**Pull request merged**